index.js

```diff
@@ -0,0 +1,101 @@
+const { PrismaClient } = require('@prisma/client')
+const { v4: uuidv4 } = require('uuid')
+
+const prisma = new PrismaClient()
+
+// retryTxn wraps SQL operations in an explicit transaction.
+// If the transaction fails, the function retries the operation, with exponential backoff.
```

Comment:
A Prisma native way to implement this could be a middleware for example. Surprisingly that almost never comes up though, so does not exist in a nice version yet. 

What would the error states be that would be solved via a retry?

Reply:
Thanks, that is interesting context. We could probably indeed do a suggested middleware for that if that is something Cockroach users would expect Prisma to handle in a nice way. (Although there might actually be a challenge about getting the right error code. Is there a way to _force_ such an error?)

Reply:
Right, the retry is appropriate if and only if the error code is exactly 40001. 

It should be possible to deterministically generate a real error of this type by beginning one transaction, committing a conflicting transaction, and attempting to commit the first transaction.

In my opinion, the default user experience should be for the client to automatically retry a few times before giving up. Middleware might be a good solution, since power users might want to customize retry timing, which they could do by replacing the middleware. Whatever solution we decide to use for this with CockroachDB should probably also be used for Prisma's Postgres experience, since the 40001 error in Postgres has the [exact same semantics](https://www.postgresql.org/docs/current/transaction-iso.html).

Reply:
Prisma currently has the stance (by default) to not retry any queries, ever and let the user implement their own control flow using middlewares. That way they have full control. On the other hand, this topic honestly almost never came up yet - especially not in context of `40001`. Surprising if you think about it...

Reply:
Now tracked at https://github.com/prisma/prisma/issues/11183 on our side.